source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+      Teuchos::RCP<MapType> columns =
+        Utilities::Trilinos::internal::make_rcp<MapType>(
+          TrilinosWrappers::types::int_type(sp.n_cols()),
+          0,
+          Utilities::Trilinos::tpetra_comm_self());
+
+      reinit_sp(
+        rows, columns, sp, false, column_space_map, graph, nonlocal_graph);
+    }
+
+
+
+    template <typename Node>
+    void
+    SparsityPattern<Node>::clear()
+    {
+      SparsityPatternBase::resize(0, 0);
+      // When we clear the matrix, reset
+      // the pointer and generate an
+      // empty sparsity pattern.
+      column_space_map = Utilities::Trilinos::internal::make_rcp<MapType>(
+        TrilinosWrappers::types::int_type(0),
+        TrilinosWrappers::types::int_type(0),
+        Utilities::Trilinos::tpetra_comm_self());
+      graph =
+        Utilities::Trilinos::internal::make_rcp<GraphType>(column_space_map,
+                                                           column_space_map,
+                                                           0);
+      graph->fillComplete();
+
+      nonlocal_graph.reset();
+    }
+
+
+
+    template <typename Node>
+    void
+    SparsityPattern<Node>::compress()
+    {
+      Assert(column_space_map.get(), ExcInternalError());
+      if (nonlocal_graph.get() != nullptr)
+        {
+          if (nonlocal_graph->getRowMap()->getLocalNumElements() > 0 &&
+              column_space_map->getGlobalNumElements() > 0)
+            {
+              // Insert dummy element at (row, column) that corresponds to row 0
+              // in local index counting.
+              TrilinosWrappers::types::int_type row =
+                nonlocal_graph->getRowMap()->getGlobalElement(0);
+              TrilinosWrappers::types::int_type column = 0;
+
+              // in case we have a square sparsity pattern, add the entry on the
+              // diagonal
+              if (column_space_map->getGlobalNumElements() ==
+                  graph->getRangeMap()->getGlobalNumElements())
+                column = row;
+              // if not, take a column index that we have ourselves since we
+              // know for sure it is there (and it will not create spurious
+              // messages to many ranks like putting index 0 on many processors)
+              else if (column_space_map->getLocalNumElements() > 0)
+                column = column_space_map->getGlobalElement(0);
+              nonlocal_graph->insertGlobalIndices(row, 1, &column);
+            }
+          Assert(nonlocal_graph->getRowMap()->getLocalNumElements() == 0 ||
+                   column_space_map->getGlobalNumElements() == 0,
+                 ExcInternalError());
+
+          nonlocal_graph->fillComplete(column_space_map, graph->getRangeMap());
+          Tpetra::Export<int, dealii::types::signed_global_dof_index, Node>
+            exporter(nonlocal_graph->getRowMap(), graph->getRowMap());
+          graph->fillComplete(column_space_map, graph->getRangeMap());
```

Comment:
You don't use the `exporter` variable. Does its constructor do all the work? If so, can you add a comment?